pandas/tools/pivot.py

```diff
@@ -136,8 +136,14 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
 
     table = agged
     if table.index.nlevels > 1:
-        to_unstack = [agged.index.names[i] or i
-                      for i in range(len(index), len(keys))]
+        index_names = agged.index.names[:len(index)]
```

Comment:
I tried this originally but some tests were failing because the MultiIndex names can also be ints. Looks like unstack first prioritizes index name if they aren't None and then defaults to index position by integer

```
In [19]: index = pd.MultiIndex.from_tuples([('one', 'a'), ('one', 'b'),
    ...: ...                                    ('two', 'a'), ('two', 'b'
    ...: )])

In [20]: s = pd.Series(np.arange(1.0, 5.0), index=index)

In [21]: s
Out[21]:
one  a    1.0
     b    2.0
two  a    3.0
     b    4.0
dtype: float64

In [22]: s.unstack(0)
Out[22]:
   one  two
a  1.0  3.0
b  2.0  4.0

In [24]: s.index.names= [1,0]

In [25]: s.unstack(0)
Out[25]:
0      a    b
1
one  1.0  2.0
two  3.0  4.0
```